cppabi32/cppabi32.rst

```diff
@@ -0,0 +1,1698 @@
+array is also 8-byte aligned.
+
+Library helper functions
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The generic C++ ABI contains some helper functions for array
+construction and destruction:
+
+.. code-block::
+
+  __cxa_vec_new     __cxa_vec_new2 
+  __cxa_vec_new3    __cxa_vec_ctor
+  __cxa_vec_dtor    __cxa_vec_cleanup
+  __cxa_vec_delete  __cxa_vec_delete2
+  __cxa_vec_delete3 __cxa_vec_cctor
+
+Compilers are not required to use these helper functions but runtime
+libraries must supply them and they must work with the always 8-byte
+cookies. These functions take pointers to constructors or destructors.
+Since constructors and destructors conforming to this ABI return *this*
+(`Summary of differences from and additions to the generic C++ ABI`_, ¶§3.1.5 *Constructor return values*, above) the return types of
+these parameters are void\* instead of void.
+
+The generic C++ ABI gives \_\_cxa\_vec\_ctor and \_\_cxa\_vec\_cctor a
+void return type. This ABI specifies void\* instead. The value returned
+is the same as the first parameter – a pointer to the array being
+constructed. We do not change the return type for \_\_cxa\_vec\_dtor
+because we provide \_\_aeabi\_vec\_dtor which has the additional
+advantage of not taking a padding\_size parameter.
+
+In addition, we define the following new helpers which can be called
+more efficiently.
+
+.. code-block::
+
+      __aeabi_vec_ctor_nocookie_nodtor
+      __aeabi_vec_ctor_cookie_nodtor
+      __aeabi_vec_cctor_nocookie_nodtor
+      __aeabi_vec_new_cookie_noctor
+      __aeabi_vec_new_nocookie
+      __aeabi_vec_new_cookie_nodtor
+      __aeabi_vec_new_cookie
+      __aeabi_vec_dtor
+      __aeabi_vec_dtor_cookie
+      __aeabi_vec_delete
+      __aeabi_vec_delete3
+      __aeabi_vec_delete3_nodtor
+      __aeabi_atexit
+
+Again, compilers are not required to use these functions but runtime
+libraries must supply them.
+
+\_\_aeabi\_vec\_dtor effectively makes \_\_cxa\_vec\_dtor obsolete.
+
+Compilers are encouraged to use the \_\_aeabi\_vec\_dtor instead of
+\_\_cxa\_vec\_dtor and \_\_aeabi\_vec\_delete instead of
+\_\_cxa\_vec\_delete. Run-time environments are encouraged to expect
+this, perhaps implementing \_\_cxa\_vec\_delete in terms of
+\_\_aeabi\_vec\_delete instead of the other way around.
+
+We define \_\_aeabi\_vec\_delete3 but not a corresponding
+\_\_aeabi\_vec\_delete2. Using that would be less efficient than using
+\_\_aeabi\_vec\_dtor and calling the ``T1::operator delete[]`` directly. See
+note 3 on page 18, below.
+
+\_\_cxa\_vec\_ctor still has uses not covered by
+\_\_aeabi\_vec\_ctor\_nocookie\_nodtor and
+\_\_aeabi\_vec\_ctor\_cookie\_nodtor.
+
+Additional helpers for array construction (i.e. new T[n],
+\_\_aeabi\_vec\_new\_\*) may be added in future releases of this ABI.
+
+Definitions of the \_\_aeabi\_\* functions are given below in terms of
+example implementations. It is not required to implement them this way.
+
+.. code-block:: c++
```

Comment:
I've added and removed blank lines as you suggest, to group the comments with their functions more sensibly.

I have an idea that some of the weird page breaking was to do with the lines being slightly too long, so that the text had to be shrunk to fit, and every so often, rst2pdf had the idea that it could end one box and start another in which the font could be larger. I think it's better to keep the font size consistent, so I've worked around it by adding a style that makes it 8pt instead of 9pt.

Still don't know of a way to get the page breaks to happen between functions, but at least the rest of the ugliness seems to have gone.

Reply:
Ah yes, that looks a lot better. Nice find :)